src/backend/distributed/commands/cascade_citus_table_function.c

```diff
@@ -0,0 +1,336 @@
+/*-------------------------------------------------------------------------
+ *
+ * cascade_citus_table_function.c
+ *   Routines to execute citus table functions (e.g undistribute_table,
+ *   create_citus_local_table) by cascading to foreign key connected
+ *   relations.
+ *
+ * Copyright (c) Citus Data, Inc.
+ *
+ *-------------------------------------------------------------------------
+ */
+
+#include "postgres.h"
+
+#include "access/xact.h"
+#include "catalog/pg_constraint.h"
+#include "distributed/commands/utility_hook.h"
+#include "distributed/commands.h"
+#include "distributed/foreign_key_relationship.h"
+#include "distributed/listutils.h"
+#include "distributed/multi_executor.h"
+#include "distributed/multi_partitioning_utils.h"
+#include "distributed/reference_table_utils.h"
+#include "distributed/relation_access_tracking.h"
+#include "distributed/worker_protocol.h"
+#include "utils/builtins.h"
+#include "utils/lsyscache.h"
+
+
+typedef void (*CitusTableFunc)(Oid, bool);
+
+
+static void EnsureSequentialModeForCitusTableCascadeFunction(void);
+static void LockRelationsWithLockMode(List *relationIdList, LOCKMODE lockMode);
+static List * FilterNonPartitionRelationIds(List *relationIdList);
+static List * GetFKeyCreationCommandsForRelationIdList(List *relationIdList);
+static void DropRelationIdListForeignKeys(List *relationIdList);
+static void DropRelationForeignKeys(Oid relationId);
+static List * GetRelationDropFkeyCommands(Oid relationId);
+static char * GetDropFkeyCascadeCommand(Oid relationId, Oid foreignKeyId);
+static void ExecuteCitusTableFunctionForRelationIdList(List *relationIdList,
+													   ExecuteCitusTableFunctionFlag
+													   executeCitusTableFunctionFlag);
+static CitusTableFunc GetCitusTableFunction(ExecuteCitusTableFunctionFlag
+											executeCitusTableFunctionFlag);
+
+
+/*
+ * ExecuteCitusTableFunctionCascade executes citus table function specified
+ * by ExecuteCitusTableFunctionFlag argument on each relation that relation
+ * with relationId is connected via it's foreign key graph, which includes
+ * input relation itself.
+ * Also see ExecuteCitusTableFunctionFlag enum definition for supported
+ * citus table functions.
+ */
+void
+ExecuteCitusTableFunctionCascade(Oid relationId, LOCKMODE lockMode,
+								 ExecuteCitusTableFunctionFlag
+								 executeCitusTableFunctionFlag)
+{
+	/*
+	 * As we will operate on foreign key connected relations, here we
+	 * invalidate foreign key graph to be on the safe side.
+	 */
+	InvalidateForeignKeyGraph();
```

Comment:
adding this and changing base branch to master

Reply:
In order to lock proper relations, we first need to have a fresh graph.
Also, `GetForeignKeyConnectedRelationIdList` already acquires locks on relations when building graph, so I think we should be safe with current approach.